diff --git a/src/spandsp/unaligned.h b/src/spandsp/unaligned.h
index 189b870..74b789f 100644
--- a/src/spandsp/unaligned.h
+++ b/src/spandsp/unaligned.h
@@ -28,6 +28,9 @@
 #if !defined(_SPANDSP_UNALIGNED_H_)
 #define _SPANDSP_UNALIGNED_H_
 
+#define __USE_MISC 1
+#include <endian.h>
+
 #if defined(__cplusplus)
 extern "C"
 {
@@ -41,9 +44,9 @@ extern "C"
 #define PACK( __Declaration__ ) __pragma( pack(push, 1) ) __Declaration__ __pragma( pack(pop))
 #endif
 
-PACK(struct __dealign_uint16 { uint16_t datum; };)
-PACK(struct __dealign_uint32 { uint32_t datum; };)
-PACK(struct __dealign_uint64 { uint64_t datum; };)
+PACK(struct __dealign_uint16 { uint16_t datum; });
+PACK(struct __dealign_uint32 { uint32_t datum; });
+PACK(struct __dealign_uint64 { uint64_t datum; });
 
 #if defined(__GNUC__)
 
